g2p/mappings/langs/network_lite.py

```diff
@@ -0,0 +1,164 @@
+            self._edges[u] = []
+
+    def add_edge(self, u, v):
+        """Add a directed edge from u to v"""
+        self.add_node(u)
+        self.add_node(v)
+        if v not in self._edges[u]:
+            self._edges[u].append(v)
+
+    def add_edges_from(self, edges: Iterable):
+        """Add edges from a list of tuples"""
+        for u, v in edges:
+            self.add_edge(u, v)
+
+    @property  # read-only
+    def nodes(self):
+        """Return the nodes"""
+        return self._edges.keys()
+
+    @property  # read-only
+    def edges(self) -> Iterator:
+        """Iterate over all edges"""
+        for u, neighbours in self._edges.items():
+            for v in neighbours:
+                yield u, v
+
+    def __contains__(self, u) -> bool:
+        """Check if a node is in the graph"""
+        return u in self._edges
+
+    def has_path(self, u, v) -> bool:
+        """Check if there is a path from u to v"""
+        if v not in self._edges:
+            raise KeyError(f"Node {v} not in graph")
+        visited: set = set()
+        return self._has_path(u, v, visited)
+
+    def _has_path(self, u, v, visited: set) -> bool:
+        """Helper function for has_path"""
+        visited.add(u)
+        if u == v:
+            return True
+        for neighbour in self._edges[u]:
+            if neighbour not in visited:
+                if self._has_path(neighbour, v, visited):
+                    return True
+        return False
+
+    def successors(self, u) -> Iterator:
+        """Return the successors of u"""
+        return iter(self._edges[u])
+
+    def descendants(self, u) -> set:
+        """Return the descendants of u"""
+        visited: set = set()
+        self._descendants(u, visited)
+        visited.remove(u)
+        return visited
+
+    def _descendants(self, u, visited: set):
+        """Helper function for descendants"""
+        visited.add(u)
+        for neighbour in self._edges[u]:
+            if neighbour not in visited:
+                self._descendants(neighbour, visited)
+
+    def ancestors(self, u):
+        """Return the ancestors of u"""
+        reversed_graph = DiGraph()
+        reversed_graph.add_edges_from((v, u) for u, v in self.edges)
+        for node in self.nodes:
+            reversed_graph.add_node(node)
+        return reversed_graph.descendants(u)
+
+    def shortest_path(self, u, v) -> list:
+        """Return the shortest path from u to v
+
+        Algorithm: Dijsktra's algorithm for unweighted graphs, which is just BFS
+
+        Returns:
+            list: the shortest path from u to v
+        Raises:
+            KeyError: if u or v is not in the graph
+            ValueError: if there is no path from u to v
+        """
+
+        if v not in self._edges:
+            raise KeyError(f"Node {v} not in graph")
+        visited = {u: None}  # dict of {node: predecessor on shortest path from u}
+        queue: deque[Any] = deque()
+        while True:
```

Comment:
This can just be `while queue`.  If it terminates without finding a path then raise ValueError.  I have written versions of this function about a zillion times, e.g. https://github.com/dhdaines/pdfplumber/blob/develop/pdfplumber/structure.py#L125

Reply:
I almost wrote it with `while queue`, but I chose to not do `queue.append(u)` before the loop starts, and exit in the second last statement of the loop instead of in the while condition.

It's the same algorithm, I know, but I have a mild preference for the way I wrote it, which avoids pushing `u` and then immediate popping it as the loop starts. Trivial and pointless optimization, I'm aware... but it felt right.

Reply:
Ah, usually I do `queue = deque([u])` ... it makes the code shorter...